api/allocation_api.py

```diff
@@ -0,0 +1,123 @@
+from rest_framework import permissions, serializers, viewsets
+from rest_framework.exceptions import ValidationError
+
+from allocation.allocation_data_builder import AllocationDataBuilder
+from allocation.allocation_solver import AllocationSolver
+from allocation.models import AllocationRequest
+from api.applications_api.serializers import NullableCurrentUserDefault
+from applications.models import (
+    ApplicationRound,
+    ApplicationRoundBasket,
+    ApplicationRoundStatus,
+)
+from permissions.api_permissions import AllocationRequestPermission
+
+
+class AllocationInProcessException(Exception):
+    pass
+
+
+class UnsuitableStatusForAllocationException(Exception):
+    pass
+
+
+class AllocationRequestSerializer(serializers.ModelSerializer):
+    start_date = serializers.DateTimeField(read_only=True)
+    end_date = serializers.DateTimeField(read_only=True)
+    completed = serializers.BooleanField(read_only=True)
+    user = serializers.HiddenField(default=NullableCurrentUserDefault())
+    application_round_id = serializers.IntegerField(required=True)
+    application_round_basket_ids = serializers.PrimaryKeyRelatedField(
+        queryset=ApplicationRoundBasket.objects.all(),
+        source="application_round_baskets",
+        many=True,
+    )
+
+    class Meta:
+        model = AllocationRequest
+        fields = [
+            "id",
+            "application_round_id",
+            "start_date",
+            "end_date",
+            "completed",
+            "user",
+            "application_round_basket_ids",
+        ]
+
+    def create(self, validated_data):
+        validated_data["start_date"] = datetime.now()
+
+        application_round_id = validated_data["application_round_id"]
+
+        application_round = ApplicationRound.objects.get(pk=application_round_id)
+
+        if application_round.get_status().status not in [
+            ApplicationRoundStatus.REVIEW_DONE,
+            ApplicationRoundStatus.ALLOCATED,
+        ]:
+            raise ValidationError(
+                f"Can only allocate application rounds when "
+                f"{ApplicationRoundStatus.REVIEW_DONE} or {ApplicationRoundStatus.ALLOCATED}"
+            )
+
+        matching_requests = AllocationRequest.objects.filter(
+            application_round__id=application_round_id, end_date=None
+        )
+
+        if len(matching_requests) > 0:
+            raise ValidationError(
+                f"Allocation in process for application round {application_round_id}."
+            )
+
+        allocation_request = super().create(validated_data)
+        t = threading.Thread(
+            target=self.start_allocation, args=[allocation_request], daemon=True
+        )
+        t.start()
+        return allocation_request
+
+    def update(self, instance, validated_data):
+        return AllocationRequest.objects.get(pk=instance.id)
+
+    def start_allocation(self, allocation_request: AllocationRequest):
```

Comment:
Should we put logic in this method under some facade method to allocation code api?

Reply:
Yeah, i think that'd probably be better place and not pollute serializer with this.